pkg/logql/expr.y

```diff
@@ -7,35 +7,81 @@ import (
 %}
 
 %union{
-  Expr         Expr
-  Filter       labels.MatchType
-  Selector     []*labels.Matcher
-  Matchers     []*labels.Matcher
-  Matcher      *labels.Matcher
-  str          string
-  int          int64
+  Expr                    Expr
+  LogExpr                 LogExpr
+  RangeAggregationExpr    SampleExpr
+  VectorAggregationExpr   SampleExpr
+  LogRangeExpr            *logRange
+  Filter                  labels.MatchType
+  Selector                []*labels.Matcher
+  Matchers                []*labels.Matcher
+  Matcher                 *labels.Matcher
+  Grouping                *grouping
+  Labels                  []string
+  VectorOp                string
+  str                     string
+  int                     int64
 }
 
 %start root
 
-%type <Expr>         expr
-%type <Filter>       filter
-%type <Selector>     selector
-%type <Matchers>     matchers
-%type <Matcher>      matcher
+%type <Expr>                  expr
+%type <Filter>                filter
+%type <Selector>              selector
+%type <Matchers>              matchers
+%type <Matcher>               matcher
+%type <VectorOp>              vectorOp
+%type <Labels>                labels
+%type <Grouping>              grouping
+%type <LogExpr>               logExpr
+%type <RangeAggregationExpr>  rangeAggregationExpr
+%type <VectorAggregationExpr> vectorAggregationExpr
+%type <LogRangeExpr>          logRangeExpr
 
 %token <str>  IDENTIFIER STRING
-%token <val>  MATCHERS LABELS EQ NEQ RE NRE OPEN_BRACE CLOSE_BRACE COMMA DOT PIPE_MATCH PIPE_EXACT
+%token <val>  MATCHERS LABELS EQ NEQ RE NRE OPEN_BRACE CLOSE_BRACE OPEN_BRACKET CLOSE_BRACKET COMMA DOT PIPE_MATCH PIPE_EXACT 
+              OPEN_PARENTHESIS CLOSE_PARENTHESIS BY WITHOUT COUNT_OVER_TIME RATE SUM AVG MAX MIN COUNT STDDEV STDVAR BOTTOMK TOPK
 
 %%
 
 root: expr { exprlex.(*lexer).expr = $1 };
 
 expr:
-      selector                         { $$ = &matchersExpr{ matchers: $1 } }
-    | expr filter STRING               { $$ = NewFilterExpr( $1, $2, $3 ) }
-    | expr filter error
-    | expr error
+      logExpr                    { $$ = $1 }
+    | rangeAggregationExpr       { $$ = $1 }
+    | vectorAggregationExpr      { $$ = $1 }
+    ;
+
+logExpr:
+      selector                         { $$ = newMatcherExpr($1)}
+    | logExpr filter STRING            { $$ = NewFilterExpr( $1, $2, $3 ) }
+    | logExpr filter error
+    | logExpr error
+    ;
+
+
+logRangeExpr:
+      selector OPEN_BRACKET IDENTIFIER IDENTIFIER CLOSE_BRACKET                                    { $$ = mustNewRange(newMatcherExpr($1), $3+$4) }
+    | OPEN_PARENTHESIS logExpr CLOSE_PARENTHESIS OPEN_BRACKET IDENTIFIER IDENTIFIER CLOSE_BRACKET  { $$ = mustNewRange($2, $5+$6) }
+    ;
```

Comment:
This seems suboptimal.  Can we make the lexer correctly identify `1m` etc?

Also, this rule doesn't feel right to me; hard to put my finger on why, but I always expect to see these expressions be like:

```yacc
expr:
| int
| expr + expr
| ( expr )
;
```

And not see the bracketing handled in a separate rule.

Reply:
The `IDENTIFIER IDENTIFIER` is caused by the go lexer, I'll give a try to improve it.

For the second point I'm trying to parse a log range which has 2 terminals:

- {foo="bar"}[5m]
- ({foo="bar"} | regex )[5m]

This is why is declared like that.

Reply:
Fixed the lexer and the logRangeExpr.

Please let me know if anything needs adjustment in the yacc files, I'm a beginner here.